be/src/exec/vectorized/partition/partition_hash_map.h

```diff
@@ -0,0 +1,288 @@
+        for (auto& [key, value] : hash_map) {
+            flush(value, chunk);
+        }
+    }
+
+    // Append chunk to the hash_map for one nullable key
+    // We maintain an additional PartitionChunks called null_key_value for null key
+    // Each key mapped to a PartitionChunks which holds an array of chunks
+    // chunk will be divided into multiply parts by partition key, and each parts will be append to the
+    // last chunk of PartitionChunks.chunks. New chunk will be allocated if the last chunk reaches its capacity
+    // @key_loader used to load key for number type or slice type
+    template <typename HashMap, typename KeyLoader, typename KeyAllocator>
+    void append_chunk_for_one_nullable_key(HashMap& hash_map, PartitionChunks& null_key_value, ChunkPtr chunk,
+                                           const NullableColumn* nullable_key_column, KeyLoader&& key_loader,
+                                           KeyAllocator&& key_allocator) {
+        if (nullable_key_column->only_null()) {
+            const auto size = chunk->num_rows();
+            if (null_key_value.chunks.empty() || null_key_value.remain_size <= 0) {
+                if (!null_key_value.chunks.empty()) {
+                    null_key_value.chunks.back()->append_selective(*chunk, null_key_value.select_indexes.data(), 0,
+                                                                   null_key_value.select_indexes.size());
+                }
+
+                alloc_new_buffer(null_key_value, chunk);
+            }
+            int32_t offset = 0;
+            auto cur_remain_size = size;
+            while (null_key_value.remain_size < cur_remain_size) {
+                null_key_value.chunks.back()->append(*chunk, offset, null_key_value.remain_size);
+                offset += null_key_value.remain_size;
+                cur_remain_size -= null_key_value.remain_size;
+
+                alloc_new_buffer(null_key_value, chunk);
+            }
+            null_key_value.chunks.back()->append(*chunk, offset, cur_remain_size);
+            null_key_value.remain_size = chunk_size - null_key_value.chunks.back()->num_rows();
+        } else {
+            const auto& null_flag_data = nullable_key_column->null_column()->get_data();
+            const auto size = chunk->num_rows();
+
+            for (uint32_t i = 0; i < size; i++) {
+                PartitionChunks* value_ptr = nullptr;
+                if (null_flag_data[i] == 1) {
+                    value_ptr = &null_key_value;
+                } else {
+                    const auto& key = key_loader(i);
+                    auto iter = hash_map.lazy_emplace(
+                            key, [&](const auto& ctor) { return ctor(key_allocator(key), PartitionChunks()); });
+                    value_ptr = &iter->second;
+                }
+
+                auto& value = *value_ptr;
+                if (value.chunks.empty() || value.remain_size <= 0) {
+                    if (!value.chunks.empty()) {
```

Comment:
if value_chunks.empty(), then alloc_new_buffer, and then push back index `i`?

Reply:
yes

Reply:
but you have not execute append_selective